lib/ansible/modules/files/find.py

```diff
@@ -456,7 +450,12 @@ def main():
                                 r['checksum'] = module.sha1(fsname)
                             filelist.append(r)
 
-                    elif stat.S_ISLNK(st.st_mode) and params['file_type'] == 'link':
+                    elif ((stat.S_ISBLK(st.st_mode) and params['file_type'] == 'block') or
+                          (stat.S_ISCHR(st.st_mode) and params['file_type'] == 'char') or
+                          (stat.S_ISDIR(st.st_mode) and params['file_type'] == 'directory') or
+                          (stat.S_ISFIFO(st.st_mode) and params['file_type'] == 'fifo') or
+                          (stat.S_ISLNK(st.st_mode) and params['file_type'] == 'link') or
+                          (stat.S_ISSOCK(st.st_mode) and params['file_type'] == 'socket')):
```

Comment:
I think an internal map with the file_type to 'stat operation' matching would be a cleaner way to implement this and then have:

```statop = getattr(stat, FILE_MAP[params['file_type']]``` to find the function for comparison